src/persist/src/indexed/background.rs

```diff
@@ -164,69 +192,162 @@ impl<B: Blob> Maintainer<B> {
+        // a ColumnarRecords
         let mut updates = vec![];
-        let mut batches = vec![];
+        let mut builder = match key_val_data_max_len {
+            None => ColumnarRecordsVecBuilder::default(),
+            Some(key_val_data_max_len) => {
+                ColumnarRecordsVecBuilder::new_with_len(key_val_data_max_len)
+            }
+        };
 
-        for key in first.keys.iter() {
-            batches.push(
-                blob.get_trace_batch_async(key, CacheHint::NeverAdd)
-                    .recv()?,
-            );
-        }
+        let mut new_size_bytes = 0;
+        let mut keys = vec![];
+
+        loop {
+            // Pull more data if either of the holding pens from first or second are empty
+            if first_updates.is_empty() {
+                if let Some(first) = first_batch_iter.next() {
+                    let first_batch = blob
+                        .get_trace_batch_async(first, CacheHint::NeverAdd)
+                        .recv()?;
+                    first_updates.extend(first_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
 
-        for key in second.keys.iter() {
-            batches.push(
-                blob.get_trace_batch_async(key, CacheHint::NeverAdd)
-                    .recv()?,
-            );
-        }
+            if second_updates.is_empty() {
+                if let Some(second) = second_batch_iter.next() {
+                    let second_batch = blob
+                        .get_trace_batch_async(second, CacheHint::NeverAdd)
+                        .recv()?;
+                    second_updates.extend(second_batch.updates.iter().flat_map(|u| {
+                        u.iter()
+                            .map(|((k, v), t, d)| ((k.to_vec(), v.to_vec()), t, d))
+                    }));
+                }
+            }
 
-        for batch in batches.iter() {
-            updates.extend(batch.updates.iter().flat_map(|u| u.iter()));
-        }
+            // Figure out the stopping point for records we can safely consolidate.
+            // This is the min max value across the two batches we have seen so far.
+            let threshold_key_val = match (first_updates.last(), second_updates.last()) {
+                (Some((f, _, _)), Some((s, _, _))) => {
+                    if f < s {
+                        f.clone()
+                    } else {
+                        s.clone()
+                    }
+                }
+                (Some((f, _, _)), None) => f.clone(),
+                (None, Some((s, _, _))) => s.clone(),
+                (None, None) => break,
+            };
 
-        for ((_, _), ts, _) in updates.iter_mut() {
-            ts.advance_by(desc.since().borrow());
-        }
+            // Drain all of the records <= threshold out of the various holding
```

Comment:
one thing that tripped me up while reviewing was trying to get a feel for the size of the output batches under various worst-case conditions. specifically, I'm not sure I see anything that puts a lower bound on the size of the parts this outputs. so the situation I'm thinking out is as batches get repeatedly compacted over time, whether anything keeps them from being chopped up in increasingly more and more parts.

an alternative is instead of flushing `updates` every time we finish one of the input batches, we repeatedly pull whichever one is necessary until `updates` hits some certain size (not just a num records, it would include key and val lens). the "what is the maximum memory usage" analysis in this case is a bit more complicated, but I _think_ it ends up as something like `first batch part size + second batch part size + 2*key_val_data_max_len`?

Reply:
So, I think there's a slight misunderstanding here. Zooming out a little bit:

`first_updates` and `second_updates` are buffers holding data from one part of the first/second trace batch respectively. We know we can stop pulling data when both are empty, and they are each bounded by the size of a `BlobTraceBatchPart`

(poorly named) `updates` is the buffer that holds the subset of `first_updates` and `second_updates` that can be forwarded to `since`, consolidated, and then placed into `builder`. We move records from `updates` -> `builder` one by one after they have been consolidated.

As we push things into builder, we only write a new `BlobTraceBatchPart` into storage once a `ColumnarRecords` has been filled, and also once at the very end in case there's leftover data that did not fill up a `ColumnarRecords`. So the lower bound on the number of batches is `(size of consolidated data) / (size limits of a ColumnarRecords) + 1` or something like that.

Does that make sense? I think we are already doing what you've suggested here - we repeatedly

1. figure out which of first_updates and second_updates is empty, and attempt to read the next batch
2. figure out the threshold of updates that can be consolidated
3. consolidate them and move them into builder
4. write out batches as the number of records in builder fills up a ColumnarRecords
5. rinse and repeat.

Let me know if this explanation isn't clear.

Reply:
let's hop on a zoom! we're definitely talking past each other, but I don't yet see exactly how

Reply:
talked about this offline. i was misunderstanding exactly what the code was doing, so this concern is a no-op. new feedback is that we might want to think a bit about how to tweak the code to make what is going on a bit more obvious

Reply:
I added a new commit that adds 3 helper functions to make what the compaction loop is doing hopefully a bit more obvious